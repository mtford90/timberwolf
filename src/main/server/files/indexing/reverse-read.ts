import fs from "mz/fs";
import { Observable } from "rxjs";

const NEW_LINE_CHARACTERS = ["\n"];

async function readPreviousChar(
  fileSize: number,
  file: number,
  currentCharacterCount: number
) {
  const bytesReadAndBuffer = await fs.read(
    file,
    Buffer.alloc(1),
    0,
    1,
    fileSize - 1 - currentCharacterCount
  );

  return String.fromCharCode((bytesReadAndBuffer as never)[1][0]);
}

export function reverseRead(
  inputFilePath: string,
  encoding: BufferEncoding = "utf8"
): Observable<{ line?: string; bytesRead: number; size: number }> {
  return new Observable<{
    line?: string;
    bytesRead: number;
    size: number;
  }>((subscriber) => {
    let file: number | null = null;

    async function read() {
      const stats = await fs.stat(inputFilePath);
      file = await fs.open(inputFilePath, "r");
      const fileSize = stats.size;

      subscriber.next({ bytesRead: 0, size: fileSize });

      let chars = 0;
      let currentLine = "";

      function nextLine() {
        subscriber.next({
          line: Buffer.from(currentLine, "binary").toString(encoding),
          bytesRead: chars,
          size: fileSize,
        });
      }

      // eslint-disable-next-line no-restricted-syntax
      while (chars < fileSize) {
        // eslint-disable-next-line no-await-in-loop
        const nextCharacter = await readPreviousChar(fileSize, file, chars);
        const nextCharIsNewLine = NEW_LINE_CHARACTERS.includes(nextCharacter);
        chars++;
        if (nextCharIsNewLine) {
          nextLine();
          currentLine = "";
        } else {
          currentLine = nextCharacter + currentLine;
        }
      }

      if (currentLine.length) {
        nextLine();
      }
    }

    const promise = read();

    promise
      .then(() => {
        subscriber.complete();
      })
      .catch((err) => {
        subscriber.error(err);
      });

    return () => {
      promise.cancel();
      if (file) {
        // noinspection JSIgnoredPromiseFromCall
        fs.close(file);
      }
    };
  });
}
