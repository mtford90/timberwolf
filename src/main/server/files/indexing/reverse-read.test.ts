import { toArray } from "rxjs/operators";
import { reverseRead } from "./reverse-read";

const path = "/Users/mtford/Playground/log/log.txt";

describe("reverse read", () => {
  describe("with events", () => {
    describe("when reading a file", () => {
      it("should read all lines in the file", async () => {
        const lines = await reverseRead(path).pipe(toArray()).toPromise();

        console.log(lines.length);

        expect(lines).toMatchSnapshot();
      });
    });
  });
});
