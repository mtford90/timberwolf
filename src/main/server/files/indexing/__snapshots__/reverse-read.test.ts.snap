// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reverse read with events when reading a file should read all lines in the file 1`] = `
Array [
  "",
  " DEBUG  [workflow-upload-orchestrator] no operations performed - sleeping",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] event was received during previous operation - trying another tick",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/error_1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073741952, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073741888, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [64]}",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
  " DEBUG  [workflow-upload-orchestrator] marked workflow folder as errored",
  " DEBUG  [workflow-upload-orchestrator] no payload file to upload",
  " DEBUG  [workflow-upload-orchestrator] no binary info files to upload",
  " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
  " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": null, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/workflow_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 512, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [512]}",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
  " DEBUG  [workflow-upload-orchestrator] deleted workflow payload",
  " DEBUG  [workflow-payload-uploader] uploadWorkflowPayload(attempt: 1): success",
  " DEBUG  [workflow-payload-uploader] uploadWorkflowPayload(attempt: 1)",
  " DEBUG  [workflow-upload-orchestrator] no binary info files to upload",
  " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
  " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] error when performing operation - sleeping for 250ms before ticking again",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/error_info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 64, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [64]}",
  " DEBUG  [workflow-upload-orchestrator] marking session binary info as errored",
  " DEBUG  [workflow-upload-orchestrator] no binaries to upload",
  " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] operation was performed - trying another tick",
  " DEBUG  [workflow-upload-orchestrator] deleted session binary",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (4 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 512, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [512]}",
  " DEBUG  [session-binary-uploader] uploadFile(attempt: 1): 45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd, (/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd) - success",
  " DEBUG  [session-binary-uploader] uploadFile(attempt: 1): 45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd, (/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd)",
  " DEBUG  [workflow-upload-orchestrator] found workflow to upload {\\"workflowToUpload\\": {\\"customerId\\": \\"ddd18590-6ab7-11e9-94af-3aaca69310c8\\", \\"error\\": false, \\"fileName\\": \\"1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"payload\\": {\\"applicationId\\": \\"c1b2bde3-52f3-44fe-9c9a-aa9c15ec9734\\", \\"applicationVersion\\": \\"0.4.1\\", \\"dateTime\\": 2020-02-17T19:15:08.836Z, \\"id\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"organizationId\\": \\"8351ec7a-6ab9-11e9-9d14-1aa083c926ce\\", \\"tz\\": \\"Europe/London\\", \\"workflowDefinitionId\\": \\"ec5f8087-822a-4967-b7e4-d1ed72ddc10a\\", \\"workflowDefinitionVersion\\": \\"0.0.5\\", \\"workflowInputs\\": [Object], \\"workflowInputsSchema\\": \\"MonitoringV1.json\\", \\"workflowTypeId\\": \\"c730ae0a-a2cd-11e7-8b2d-b2b3c9a1a96a\\"}, \\"timestamp\\": 1581966924343, \\"workflowId\\": \\"0999f04a-986e-4740-be21-2b4ffc2dedda\\"}}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (3 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/workflow_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 256, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [256]}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (2 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/info_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd.json\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
  " DEBUG  [workflow-upload-orchestrator] tick ignored - already performing operation (1 ignored events)",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda/binary_45c8e1b0-6d87-4bc1-ba90-0b2ca1d452bd\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 128, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [128]}",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " DEBUG  [workflow-upload-orchestrator] file observation {\\"file\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda\\", \\"listenerId\\": \\"9a59ed7c-eff0-42ba-aa64-5bf3b7c23412\\", \\"mask\\": 1073742080, \\"path\\": \\"/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows\\", \\"types\\": [256]}",
  " DEBUG  [workflow-file-manager] mkdir(/storage/emulated/0/Android/data/com.xpertsea.mobileapp.staging/files/workflows/a830a970-f987-11e9-b4d8-66674c84890a/ddd18590-6ab7-11e9-94af-3aaca69310c8/8351ec7a-6ab9-11e9-9d14-1aa083c926ce/1581966924343_0999f04a-986e-4740-be21-2b4ffc2dedda)",
  " DEBUG  [workflow-upload-orchestrator] no operations performed - sleeping",
  " DEBUG  [workflow-upload-orchestrator] performOperation",
  " LOG  Running \\"MobileApp\\" with {\\"rootTag\\":11}",
  "",
  " BUNDLE  [android, dev] ./index.js ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ 100.0% (2395/2395), done.",
  "Loading dependency graph, done.",
  "",
  "   /Users/mtford/Playground/xpertsea/mobile-app ",
  "Looking for JS files in",
  "",
  "└──────────────────────────────────────────────────────────────────────────────┘",
  "│                                                                              │",
  "│  https://github.com/facebook/react-native                                    │",
  "│                                                                              │",
  "│  close this tab and run your own Metro instance if you prefer.               │",
  "│  Keep Metro running while developing on any JS projects. Feel free to        │",
  "│                                                                              │",
  "│  Running Metro Bundler on port 8081.                                         │",
  "│                                                                              │",
  "┌──────────────────────────────────────────────────────────────────────────────┐",
  "$ react-native start",
  "yarn run v1.19.2",
]
`;
